library/spdm_responder_lib/measurements.c

```diff
@@ -246,266 +247,181 @@ return_status spdm_get_response_measurements(IN void *context,
 		}
 	}
 
-	device_measurement_size = sizeof(device_measurement);
+	spdm_response_size = sizeof(spdm_measurements_response_t);
+
+	signature_size = spdm_get_asym_signature_size(
+		spdm_context->connection_info.algorithm.base_asym_algo);
+
+	measurements_sig_size =
+		SPDM_NONCE_SIZE + sizeof(uint16) +
+		spdm_context->local_context.opaque_measurement_rsp_size +
+		signature_size;
+	measurements_no_sig_size =
+		SPDM_NONCE_SIZE + sizeof(uint16) +
+		spdm_context->local_context.opaque_measurement_rsp_size;
+
+	if ((spdm_request->header.param1 &
+		    SPDM_GET_MEASUREMENTS_REQUEST_ATTRIBUTES_GENERATE_SIGNATURE) !=
+		0) {
+		spdm_response_size += measurements_sig_size;
+	} else {
+		spdm_response_size += measurements_no_sig_size;
+	}
+
+	zero_mem(response, *response_size);
+
+	measurements_index = spdm_request->header.param2;
+	measurements_count = 0;
+
+	// The response buffer must hold the spdm_measurements_response_t,
+	// followed by the actual measurements, followed by the signature,
+	// if there is one. Here we calculate the maximum size allowed for
+	// measurements and store it in "measurements_size", by subtracting
+	// out "spdm_responze_size", which contains the sizeof the
+	// spdm_measurements_response_t + signature if there is one.
+
+	measurements_size = *response_size;
+	if (measurements_size > spdm_response_size) {
+		measurements_size -= spdm_response_size;
+	} else {
+		spdm_generate_error_response(spdm_context,
+					     SPDM_ERROR_CODE_UNSPECIFIED,
+					     0, response_size, response);
+		return RETURN_BUFFER_TOO_SMALL;
+	}
+
+	measurements = (uint8*)response + sizeof(spdm_measurements_response_t);
+
 	ret = spdm_measurement_collection(
 		spdm_context->connection_info.version,
 		spdm_context->connection_info.algorithm.measurement_spec,
 		spdm_context->connection_info.algorithm.measurement_hash_algo,
-		&device_measurement_count, device_measurement,
-		&device_measurement_size);
+		measurements_index,
+		&measurements_count,
+		measurements,
+		&measurements_size);
+

```

Comment:
Since spdm_measurement_collection returns boolean, there is no way to distinguish an UNSPECIFIED error and INVALID_REQUEST error (for measurements_index not being supported by the device). For the later case, I made spdm_measurement_collection return TRUE, but measurement_count == 0.